File: leetcode/number-of-islands.go
package main

import "fmt"

func numIslands(grid [][]byte) int {
	if len(grid) == 0 {
		return 0
	}
	n := len(grid)
	m := len(grid[0])
	counter := 0
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == '1' {
				counter++
				sinkIt(grid, i, j)
			}
		}
	}
	return counter
}

func sinkIt(grid [][]byte, i, j int) {
	if i < 0 || i >= len(grid) || j < 0 || j >= len(grid[0]) || grid[i][j] == '0' {
		return
	}
	grid[i][j] = '0'
	sinkIt(grid, i-1, j)
	sinkIt(grid, i+1, j)
	sinkIt(grid, i, j-1)
	sinkIt(grid, i, j+1)
}

func main() {
	inp := [][]byte{[]byte{'1', '1', '0', '0', '0'},
		[]byte{'1', '1', '0', '0', '0'},
		[]byte{'0', '0', '1', '0', '0'},
		[]byte{'0', '0', '0', '1', '1'}}
	fmt.Println(numIslands(inp))
}

File: leetcode/number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int counter  = 0;
        for(int i = 0; i < grid.size(); ++i) {
            for( int j = 0; j < grid[0].size(); ++j) {
                if (grid[i][j] == '1') {
                    counter += 1;
                    this->fill(grid, i, j);
                }
            }
        }
        return counter;
    }
    
    void fill(vector<vector<char>>& grid, int i, int j) {
        if ( i < 0 || j < 0 || i >= grid.size() || j >= grid[0].size()) { return; }
        if  (grid[i][j] == '0') { return; }
        grid[i][j] = '0';
        for (int ii = -1; ii < 2; ++ii ) {
            for (int jj = -1; jj < 2; ++jj ) {
                if (ii == 0 || jj == 0)
                  fill(grid, i+ii, j+jj );
            }
        }
        
    }
};